aten/src/ATen/native/mkldnn/RNN.cpp

```diff
@@ -0,0 +1,271 @@
+//   bias (ldgo): {num_layers, num_directions, num_biases, hidden_size}
+//
+// return vector holds weight ideep::tensor per layer per direction (num_layers = 1,
+// num_directions = 1):
+// {
+//   weight_ih_00, weight_hh_00, bias_00, /* layer = 0, direction = 0 */
+//   weight_ih_01, weight_hh_01, bias_01, /* layer = 0, direction = 1 */
+//   ..., ..., ...,
+//   weight_ih_ld, weight_hh_ld, bias_ld /* layer = l, direction = d */
+// }
+//
+std::vector<ideep::tensor> get_weight_itensors(const Tensor& flatten_weight,
+    int64_t weight_stride0, int64_t mode, int64_t input_size, int64_t hidden_size,
+    int64_t num_layers, int64_t num_directions) {
+  std::vector<ideep::tensor> weights;
+  weights.reserve(num_layers * num_directions * weight_stride0);
+
+  auto num_gates = get_num_gates(mode);
+  auto num_biases = get_num_biases(mode);
+
+  auto base = flatten_weight.data_ptr<float>();
+  int64_t offset = 0;
+  for (int64_t layer = 0; layer < num_layers; layer++) {
+    for (int64_t direction = 0; direction < num_directions; direction++) {
+      auto layer_input_size = (layer == 0) ? input_size : hidden_size * num_directions;
+      auto index = layer * num_directions + direction;
+
+      std::vector<int64_t> weight_ih_size{1, 1, layer_input_size, num_gates, hidden_size};
+      std::vector<int64_t> weight_hh_size{1, 1, hidden_size, num_gates, hidden_size};
+      std::vector<int64_t> bias_size{1, 1, num_biases, hidden_size};
+      for (std::vector<int64_t> sz : {weight_ih_size, weight_hh_size, bias_size}) {
+        weights.emplace_back(ideep::tensor{{{sz.cbegin(), sz.cend()}, dtype}, base + offset});
+        offset += std::accumulate(sz.cbegin(), sz.cend(), 1, std::multiplies<int64_t>());
+      }
+    }
+  }
+
+  return weights;
+};
+
+} // anonymous namespace
+
+Tensor _mkldnn_rnn_flatten_weight(TensorList weight, bool has_bias,
+    int64_t mode, int64_t hidden_size, int64_t num_layers, bool bidirectional) {
+  int64_t num_directions = bidirectional ? 2 : 1;
+
+  int64_t weight_stride0 = has_bias ? 4 : 2;
+  MatrixRef<Tensor> weights{weight, static_cast<size_t>(weight_stride0)};
+
+  int64_t flatten_weight_stride0 = has_bias ? 3 : 2;
+  std::vector<Tensor> flatten_weight_arr;
+  flatten_weight_arr.reserve(num_layers * num_directions * flatten_weight_stride0);
+
+  for (int64_t layer = 0; layer < num_layers; layer++) {
+    for (int64_t direction = 0; direction < num_directions; direction++) {
+      auto index = layer * num_directions + direction;
+      auto layer_weights = weights[index];
+
+      flatten_weight_arr.emplace_back(_shuffle_weight(layer_weights[0], mode).view(-1));
+      flatten_weight_arr.emplace_back(_shuffle_weight(layer_weights[1], mode).view(-1));
+      if (has_bias) {
+        flatten_weight_arr.emplace_back(_shuffle_bias(layer_weights[2], layer_weights[3], mode));
+      } else {
+        auto bias = at::zeros({get_num_biases(mode) * hidden_size});
```

Comment:
Unused `bias`

Reply:
Updated. Use zero tensor to feed into `mkl-dnn` when `bias=False`